library/Network/CloudSeeder/Types.hs

```diff
@@ -25,24 +31,33 @@ import qualified Data.Map as M
 import qualified Data.Text as T
 import qualified Data.Set as S
 
+data Stack = Stack
+  { _stackParameters :: S.Set T.Text
+  }
+  deriving (Eq, Show, Ord)
+makeFields ''Stack
+
 data ParameterSource
   = Constant T.Text -- ^ @'Constant' "param value"@
   | Env
   | Flag
   | Outputs
-  | PreviousValue
   deriving (Eq, Show, Ord)
-
 makeClassyPrisms ''ParameterSource
 
 data Parameter = Parameter ParameterSource T.Text T.Text
   deriving (Eq, Show)
 
+data ProvisionType
+  = CreateStack
+  | UpdateStack (S.Set T.Text) -- ^ @'Update' "list of parameter keys"@
+  deriving (Eq, Show, Ord)
+makeClassyPrisms ''ProvisionType
+
 data ParameterSpec
   = Required T.Text
-  | Optional T.Text T.Text
+  | Optional T.Text (Maybe T.Text)
```

Comment:
I think this is where all the `Maybe`s are coming from that I mentioned earlier. I find this a little confusing, since it’s not immediately obvious to me what a “default value of `Nothing`” *means*. I would change this one of two ways:

  1. Change `ParameterSpec` to encompass more than two cases, and give them more specific names. Something like this:

     ```hs
     data ParameterSpec
       = UserSupplied T.Text
       | TemplateDefault T.Text T.Text
       | UsePreviousValue T.Text
     ```

     You could also “denormalize” the type a little to reduce the redundancy in that every case has to carry the parameter name by splitting it into two types:

     ```hs
     data ParameterSpec = ParameterSpec
       { _parameterSpecName :: T.Text
       , _parameterSpecType :: ParameterSpecType
       }

     data ParameterSpecType
       = UserSupplied
       | TemplateDefault T.Text
       | UsePreviousValue
     ```

  2. Alternatively, you could also just make a type isomorphic to `Maybe` with more descriptive constructor names, like this:

     ```hs
     data ParameterSpec
       = Required T.Text
       | Optional T.Text ParameterDefault

     data ParameterDefault
       = TemplateDefault T.Text
       | UsePreviousValue
     ```